### PR TITLE
Procedure to compute table stats

### PR DESCRIPTION
**Pull request by karuppayya**

### CODE REVIEW

**Comment by karuppayya** [comment]
@aokolnychyi @szehon-ho Can you help review this please

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  public void removeTable() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @TestTemplate
+  public void testProcedureOnEmptyTable() {
+    sql("CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg", tableName);
+    List<Object[]> result =
+        sql("CALL %s.system.compute_table_stats('%s')", catalogName, tableIdent);
+    Assertions.assertTrue(result.isEmpty());
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:46]:
```suggestion
    assertThat(result).isEmpty();
```

please use the AssertJ assertions and we generally try to avoid ussing assertTrue/assertFalse as it doesn't provide enough context about the actual/expected when an assertion fails

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  public void testProcedureWithNamedArgs() {
+    sql(
+        "CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg PARTITIONED BY (data)",
+        tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')", tableName);
+    List<Object[]> output =
+        sql(
+            "CALL %s.system.compute_table_stats(table => '%s', columns => array('id'))",
+            catalogName, tableIdent);
+    Assertions.assertNotNull(output.get(0));
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:59]:
```suggestion
    assertThat(output.get(0)).isNotNull();
```
should this maybe also assert some more details here?

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  @TestTemplate
+  public void testProcedureWithInvalidColumns() {
+    sql(
+        "CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg PARTITIONED BY (data)",
+        tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')", tableName);
+    IllegalArgumentException illegalArgumentException =
+        Assertions.assertThrows(
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:84]: please use AssertJ's `assertThatThrownBy(..)` here. There are a bunch of examples in the codebase for this

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            IllegalArgumentException.class,
+            () ->
+                sql(
+                    "CALL %s.system.compute_table_stats(table => '%s', snapshot_id => %dL)",
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:104]: maybe also add a test with an invalid/non-existing table name

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        "CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg PARTITIONED BY (data)",
+        tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')", tableName);
+    Table table = Spark3Util.loadIcebergTable(spark, tableName);
+    Snapshot snapshot = table.currentSnapshot();
+    List<Object[]> output =
+        sql(
+            "CALL %s.system.compute_table_stats('%s', %dL)",
+            catalogName, tableIdent, snapshot.snapshotId());
+    Assertions.assertNotNull(output.get(0));
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:74]: same as above

**Comment by anton-db** [comment]
I will take a look at the partition stats PR first by @ajantha-bhat. I want to understand if we want a single analyze procedure or different procedures for table and partition stats.

**Comment by aokolnychyi** [comment]
I will take a look at the partition stats PR first by @ajantha-bhat. I want to understand if we want a single analyze procedure or different procedures for table and partition stats.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        "CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg PARTITIONED BY (data)",
+        tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')", tableName);
+    Table table = Spark3Util.loadIcebergTable(spark, tableName);
+    Snapshot snapshot = table.currentSnapshot();
+    List<Object[]> output =
+        sql(
+            "CALL %s.system.compute_table_stats('%s', %dL)",
+            catalogName, tableIdent, snapshot.snapshotId());
+    assertThat(output.get(0)).isNotNull();
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:76]: I think it would be good to check whether the output has some valid results here and in the other test

**Review by nastra — APPROVED** [review]
LGTM, but I left a few comments around testing that would be good to address

**Comment by karuppayya** [comment]
@aokolnychyi Can this merged now?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```
Comment (szehon-ho) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java:83]: curious why is this needed?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  public InternalRow[] call(InternalRow args) {
+    ProcedureInput input = new ProcedureInput(spark(), tableCatalog(), PARAMETERS, args);
+    Identifier tableIdent = toIdentifier(args.getString(0), PARAMETERS[0].name());
```
Comment (szehon-ho) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java:86]: cant we use input.ident()?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    Long snapshotId = input.isProvided(SNAPSHOT_ID_PARAM) ? input.asLong(SNAPSHOT_ID_PARAM) : null;
```
Comment (szehon-ho) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java:88]: i think we should use input API that has default value?

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    verifyTableStats(tableName);
+  }
+
+  @TestTemplate
+  public void testProcedureWithInvalidColumns() {
+    sql(
+        "CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg PARTITIONED BY (data)",
+        tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')", tableName);
```
Comment (szehon-ho) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:89]: Nit: could we remove the insert to make test shorter?  (on next test as well)

**Review by szehon-ho — APPROVED** [review]

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java (hunk 10 above)
Comment (szehon-ho) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:89]: Actually @karuppayya did we miss this comment?  Will be nice in negative case, to cut down unnecessary test runtime

Comment (karuppayya) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java]: Removed the insert

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java (hunk 9 above)
Comment (karuppayya) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java]: Changed to use the API that has default value

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java (hunk 7 above)
Comment (karuppayya) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java]: Removed. 
This was copied from a different test.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                    catalogName, tableIdent, 1234L))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Snapshot not found");
+  }
+
+  void verifyTableStats(String tableName) throws NoSuchTableException, ParseException {
+    Table table = Spark3Util.loadIcebergTable(spark, tableName);
+    StatisticsFile statisticsFile = table.statisticsFiles().get(0);
+    BlobMetadata blobMetadata = statisticsFile.blobMetadata().get(0);
+    assertThat(
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:120]: `assertThat(blobMetadata.properties()).containsEntry(...)` or just `containsKey(..)`. The advantage here is that it will show the properties if the assertion ever fails

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java (hunk 10 above)
Comment (karuppayya) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:93]: @szehon-ho I just realized that the insert was added intentionally, to create a snapshot.
The validation for columns(this test) happens only when the [table has snapshots](https://github.com/apache/iceberg/blob/main/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/ComputeTableStatsSparkAction.java#L90).

**Comment by szehon-ho** [comment]
This looks good to me, will merge tomorrow if no additional comments

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java (hunk 4 above)
Comment (nastra) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestComputeTableStatsProcedure.java:111]: @karuppayya can you please add a test with an invalid/non-existing table?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java [hunk 12]
```diff
@@ -61,6 +61,7 @@ private static Map<String, Supplier<ProcedureBuilder>> initProcedureBuilders() {
     mapBuilder.put("create_changelog_view", CreateChangelogViewProcedure::builder);
     mapBuilder.put("rewrite_position_delete_files", RewritePositionDeleteFilesProcedure::builder);
     mapBuilder.put("fast_forward", FastForwardBranchProcedure::builder);
+    mapBuilder.put("compute_table_stats", ComputeTableStatsProcedure::builder);
```
Comment (ajantha-bhat) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java:64]: Please also document it under 
https://github.com/apache/iceberg/blob/main/docs/docs/spark-procedures.md

Comment (karuppayya) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java:64]: Thanks @ajantha-bhat , 
I will create a follow up PR for the doc changes, just to keep this PR focussed on the procedure.

**Comment by aokolnychyi** [comment]
I'd love to review today as well.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    ProcedureInput input = new ProcedureInput(spark(), tableCatalog(), PARAMETERS, args);
+    Identifier tableIdent = input.ident(TABLE_PARAM);
+
```
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java:86]: Why do we have this line to separate `snapshotId` and `columns`?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        table -> {
+          ComputeTableStats action = actions().computeTableStats(table);
+          if (snapshotId != null) {
+            action.snapshot(snapshotId);
+          }
+          if (columns != null) {
+            action.columns(columns);
+          }
+
+          ComputeTableStats.Result result = action.execute();
```
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java:101]: What about a direct import for `Result` for shorter lines?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    Identifier tableIdent = input.ident(TABLE_PARAM);
+
+    Long snapshotId = input.asLong(SNAPSHOT_ID_PARAM, null);
+    String[] columns = input.asStringArray(COLUMNS_PARAM, null);
+
+    return modifyIcebergTable(
+        tableIdent,
+        table -> {
+          ComputeTableStats action = actions().computeTableStats(table);
+          if (snapshotId != null) {
```
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/ComputeTableStatsProcedure.java:94]: What about an empty line before each `if` block to separate them?

**Review by aokolnychyi — APPROVED** [review]
Minor styling suggestions. LGTM.
